File: python/gtk4/application.py
import gi, sys
gi.require_version("Gtk", "4.0")
from gi.repository import Gtk

appID = "io.github.Miqueas.GTK-Examples.Python.Gtk3.Application"

app = Gtk.Application(application_id = appID)
app.connect("activate", lambda _: print("app::activate called"))
app.connect("startup", lambda _: print("app::startup called"))

sys.exit(app.run(sys.argv))

File: zig/gtk4/application.zig
const std = @import("std");
const gtk = @import("gtk");
const gio = @import("gio");

const APP_ID = "io.github.Miqueas.GTK-Examples.Zig.Application";

pub fn main(init: std.process.Init) void {
    var app = gtk.Application.new(APP_ID, .{});
    defer app.unref();

    _ = gio.Application.signals.startup.connect(app, ?*anyopaque, &onStartup, null, .{});
    _ = gio.Application.signals.activate.connect(app, ?*anyopaque, &onActivate, null, .{});

    const argv = init.minimal.args.vector;
    const status = app.as(gio.Application).run(
        @intCast(argv.len),
        @ptrCast(@constCast(argv.ptr))
    );

    std.process.exit(@intCast(status));
}

fn onActivate(_: *gtk.Application, _: ?*anyopaque) callconv(.c) void {
    std.debug.print("Hello there!\n", .{});
}

fn onStartup(_: *gtk.Application, _: ?*anyopaque) callconv(.c) void {
    std.debug.print("Initializing app... ", .{});
    std.debug.print("Done!\n", .{});
}